### PR TITLE
replace OutBuffer.reset() with .setsize(0)

### PR DESCRIPTION
**Pull request by WalterBright**
Remove redundant functionality.

### CODE REVIEW

File: src/dmd/root/outbuffer.d [hunk 1]
```diff
@@ -62,8 +67,14 @@ struct OutBuffer
         }
     }
 
+    /************************
+     * Shrink the size of the data to `size`.
+     * Params:
+     *  size = new size of data, must be <= `.length`
+     */
     extern (C++) void setsize(size_t size) pure nothrow @nogc @safe
```
Comment (marler8997) [inline src/dmd/root/outbuffer.d]: would this be a good time to rename it to `shrink`?

**Comment by rainers** [comment]
Not a fan of this, `reset` is used so often and is more obvious to read IMO. It also has a smaller call overhead though that might disappear with inlining anyway.

**Comment by rainers** [comment]
> Had to add `fillVoid()` so usage could be made memory safe.

Looks good, but seems like it's unrelated to `reset`, i.e. better as a separate PR.

File: src/dmd/dmangle.d [hunk 2]
```diff
@@ -1027,15 +1027,15 @@ public:
         buf.writeByte(m);
         buf.print(q.length);
         buf.writeByte('_');    // nbytes <= 11
-        size_t qi = 0;
-        for (char* p = cast(char*)(*buf)[].ptr + buf.length, pend = p + 2 * q.length; p < pend; p += 2, ++qi)
+        const len = buf.length;
+        auto slice = buf.fillVoid(2 * q.length);
```
Comment (rainers) [inline src/dmd/dmangle.d]: This is now reserving twice (already in line 1026, which could be omitted).

Comment (WalterBright) [reply src/dmd/dmangle.d]: It's doing something fundamentally different from reserving. The space past `offset` should be off limits to users of `OutBuffer`. `fillVoid()` actually allocates that space to the user.

**Comment by WalterBright** [comment]
> seems like it's unrelated to reset, i.e. better as a separate PR.

It is related. The `assert()` I added to `setsize()` tripped, and `fillVoid()` is the fix.

**Comment by WalterBright** [comment]
> Not a fan of this, reset is used so often and is more obvious to read IMO. It also has a smaller call overhead though that might disappear with inlining anyway.

It does disappear with inlining, and my reasoning was functions should not overlap, but be orthogonal.

**Comment by WalterBright** [comment]
In case anyone was wondering :-) what I'm trying to achieve with this series of PRs is a simpler OutBuffer with a memory safe interface. I'm even changing it to follow Ownership/Borrowing semantics. As dmd is transitioning to being a more manually managed program than a GC one, more people work on the code, and it gets more complex, this is becoming critically important.

File: src/dmd/dmangle.d (hunk 2 above)
Comment (rainers) [reply src/dmd/dmangle.d]: I understand. I was referring to `fillVoid` reserving internally, while this is also done manually above. Doesn't do much harm, though.

**Comment by rainers** [comment]
> It is related. The `assert()` I added to `setsize()` tripped, and `fillVoid()` is the fix.

ok.


> It does disappear with inlining, and my reasoning was functions should not overlap, but be orthogonal.

There is a lot of overlap with the write functions, too. Reduce them to just `write(const(void)[] buf)`?

File: src/dmd/root/outbuffer.d [hunk 3]
```diff
@@ -271,6 +282,20 @@ struct OutBuffer
...
     }
 
+    /**
+     * Allocate space, but leave it uninitialized.
+     * Params:
+     *  nbytes = amount to allocate
+     * Returns:
+     *  slice of the allocated space to be filled in
+     */
+    extern (D) char[] fillVoid(size_t nbytes) pure nothrow
```
Comment (rainers) [inline src/dmd/root/outbuffer.d]: Your description uses "allocate" everywhere, why call it "fill" then?

**Comment by WalterBright** [comment]
> There is a lot of overlap with the write functions, too. Reduce them to just write(const(void)[] buf)?

I agree that they can be pared down, not sure how far.

File: src/dmd/root/outbuffer.d (hunk 3 above)
Comment (WalterBright) [reply src/dmd/root/outbuffer.d]: Because of its similarity to `fillZero()`.

Comment (WalterBright) [reply src/dmd/root/outbuffer.d]: I think you're right, allocate is a better name.

**Review by RazvanN7 — APPROVED** [review]

**Pull request merged**